File: hkp-frontend/src/tests/grants.test.ts
import { afterEach, describe, expect, it, vi } from "vitest";

import {
  SecretRelease,
  allowedSecrets,
  grantKey,
  readGrantKey,
  resetSecretConsent,
  setSecretConsent,
} from "hkp-frontend/src/core/secretConsent";

/**
 * Where a remembered answer lives, and what it is remembered against.
 *
 * A grant is not a credential, but it is what stands between a board and one:
 * anything that can write one can have a secret handed over without being
 * asked. So the host keeps them, injecting a copy the page reads before any
 * board loads — provisioning starts as soon as one does, and a grant arriving a
 * moment later would be a question asked again for something already answered.
 */

/** A fresh copy of the grants module, reading whatever the host injected. */
async function loadGrants(injected: unknown) {
  vi.resetModules();
  (window as any).__HKP_GRANTS__ = injected;
  return import("hkp-frontend/src/grants");
}

const release = (over: Partial<SecretRelease> = {}): SecretRelease => ({
  boardName: "Mail",
  runtimeId: "node",
  url: "http://127.0.0.1:8080",
  aliases: ["gmail.imap"],
  ...over,
});

afterEach(() => {
  delete (window as any).__HKP_GRANTS__;
  resetSecretConsent();
});

describe("the key a grant is remembered against", () => {
  it("survives a board named like a separator", () => {
    // Why it is a JSON array and not joined text: these two are different
    // triples and must never collapse into one key.
    expect(grantKey(release({ boardName: "Mail node" }))).not.toBe(
      grantKey(release({ boardName: "Mail", runtimeId: "node x" })),
    );
  });

  it("reads back into the parts it was made from", () => {
    expect(readGrantKey(grantKey(release()))).toEqual({
      boardName: "Mail",
      runtimeId: "node",
      origin: "http://127.0.0.1:8080",
    });
  });

  it("says nothing for a key it did not make", () => {
    expect(readGrantKey("Mail node http://x")).toBeNull();
    expect(readGrantKey('["Mail","node"]')).toBeNull();
    expect(readGrantKey('["Mail","node",7]')).toBeNull();
  });

  it("names an unsaved board rather than leaving it blank", () => {
    expect(readGrantKey(grantKey(release({ boardName: "" })))?.boardName).toBe(
      "(unsaved)",
    );
  });
});

describe("reading what the host injected", () => {
  it("takes a grant it can use", async () => {
    const key = grantKey(release());
    const grants = await loadGrants({ [key]: ["gmail.imap"] });

    expect(grants.grantedAliases(key)).toEqual(["gmail.imap"]);
  });

  it("drops an entry that is not a list of names", async () => {
    const grants = await loadGrants({ a: "gmail.imap", b: [1, 2], c: [] });

    expect(grants.allGrants()).toEqual({});
  });

  it("is empty on a host that injected nothing", async () => {
    const grants = await loadGrants(undefined);

    expect(grants.allGrants()).toEqual({});
    expect(grants.grantedAliases("anything")).toEqual([]);
  });
});

describe("the store the consent gate uses", () => {
  it("skips the question for something the host already remembered", async () => {
    const key = grantKey(release());
    const grants = await loadGrants({ [key]: ["gmail.imap"] });
    const prompt = vi.fn(async () => ({ allowed: [], remember: false }));
    setSecretConsent({ prompt, grants: grants.hostGrantStore });

    expect(await allowedSecrets(release())).toEqual(["gmail.imap"]);
    expect(prompt).not.toHaveBeenCalled();
  });

  it("writes an answer somewhere the next launch will see", async () => {
    const grants = await loadGrants({});
    const grant = vi.fn();
    grants.setGrantPersist({ grant, revoke: vi.fn() });
    setSecretConsent({
      prompt: async (request) => ({ allowed: request.aliases, remember: true }),
      grants: grants.hostGrantStore,
    });

    await allowedSecrets(release());
    grants.setGrantPersist(null);

    expect(grant).toHaveBeenCalledWith(grantKey(release()), ["gmail.imap"]);
  });

  it("writes nothing for an answer that was not to be remembered", async () => {
    const grants = await loadGrants({});
    const grant = vi.fn();
    grants.setGrantPersist({ grant, revoke: vi.fn() });
    setSecretConsent({
      prompt: async (request) => ({ allowed: request.aliases, remember: false }),
      grants: grants.hostGrantStore,
    });

    await allowedSecrets(release());
    grants.setGrantPersist(null);

    expect(grant).not.toHaveBeenCalled();
  });

  it("adds to a grant rather than replacing it", async () => {
    // A board that comes to need `slack` is asked about `slack` alone, so the
    // answer carries only that — and must not read as withdrawing `gmail.imap`.
    const key = grantKey(release());
    const grants = await loadGrants({ [key]: ["gmail.imap"] });
    setSecretConsent({
      prompt: async (request) => ({ allowed: request.aliases, remember: true }),
      grants: grants.hostGrantStore,
    });

    await allowedSecrets(release({ aliases: ["gmail.imap", "slack"] }));

    expect(grants.grantedAliases(key)).toEqual(["gmail.imap", "slack"]);
  });

  it("forgets one, so the next release asks again", async () => {
    const key = grantKey(release());
    const grants = await loadGrants({ [key]: ["gmail.imap"] });
    const prompt = vi.fn(async () => ({ allowed: [], remember: false }));
    setSecretConsent({ prompt, grants: grants.hostGrantStore });

    grants.forgetGrant(key);

    expect(await allowedSecrets(release())).toEqual([]);
    expect(prompt).toHaveBeenCalledTimes(1);
  });

  it("does not follow the same runtime id to another address", async () => {
    const grants = await loadGrants({ [grantKey(release())]: ["gmail.imap"] });
    const prompt = vi.fn(async () => ({ allowed: [], remember: false }));
    setSecretConsent({ prompt, grants: grants.hostGrantStore });

    expect(
      await allowedSecrets(release({ url: "https://evil.example" })),
    ).toEqual([]);
    expect(prompt).toHaveBeenCalledTimes(1);
  });
});

/**
 * The store a host that keeps nothing falls back to.
 *
 * This is what the website runs: no injection, so no host store, so grants live
 * in the browser alongside everything else the tab remembers. It is the weaker
 * position — site data is cleared more casually than a file in a home
 * directory, and anything running in the page can write one — but it is what a
 * browser has, and being asked again is the direction it fails in.
 */
describe("the browser-storage fallback", () => {
  const stored = (request: SecretRelease) =>
    window.localStorage.getItem(`hkp.secret-grants:${grantKey(request)}`);

  afterEach(() => window.localStorage.clear());

  it("is what a host that registered nothing gets", async () => {
    // Nothing is registered here on purpose: the default has to be the store,
    // not an absence that quietly forgets every answer.
    setSecretConsent({
      prompt: async (r) => ({ allowed: r.aliases, remember: true }),
    });

    await allowedSecrets(release());

    expect(JSON.parse(stored(release())!)).toEqual(["gmail.imap"]);
  });

  it("does not ask a second time for what it remembered", async () => {
    const prompt = vi.fn(async (r: SecretRelease) => ({
      allowed: r.aliases,
      remember: true,
    }));
    setSecretConsent({ prompt });

    await allowedSecrets(release());
    expect(await allowedSecrets(release())).toEqual(["gmail.imap"]);
    expect(prompt).toHaveBeenCalledTimes(1);
  });

  it("adds to a grant rather than replacing it", async () => {
    setSecretConsent({
      prompt: async (r) => ({ allowed: r.aliases, remember: true }),
    });

    await allowedSecrets(release());
    await allowedSecrets(release({ aliases: ["gmail.imap", "slack"] }));

    expect(JSON.parse(stored(release())!)).toEqual(["gmail.imap", "slack"]);
  });

  it("writes nothing for an answer that was not to be remembered", async () => {
    setSecretConsent({
      prompt: async (r) => ({ allowed: r.aliases, remember: false }),
    });

    await allowedSecrets(release());

    expect(stored(release())).toBeNull();
  });

  it("reads nothing out of a key holding something else", async () => {
    window.localStorage.setItem(
      `hkp.secret-grants:${grantKey(release())}`,
      '"not a list"',
    );
    const prompt = vi.fn(async () => ({ allowed: [], remember: false }));
    setSecretConsent({ prompt });

    expect(await allowedSecrets(release())).toEqual([]);
    expect(prompt).toHaveBeenCalledTimes(1);
  });

  it("asks again rather than throwing where there is no storage", async () => {
    // A private window, or storage switched off. Both reads and writes throw,
    // and neither may take the page down with them. Replaced outright rather
    // than spied on: the one jsdom provides is a proxy a spy does not stick to.
    const real = Object.getOwnPropertyDescriptor(window, "localStorage");
    const unavailable = () => {
      throw new Error("storage is disabled");
    };
    Object.defineProperty(window, "localStorage", {
      configurable: true,
      value: { getItem: unavailable, setItem: unavailable },
    });

    try {
      const prompt = vi.fn(async (r: SecretRelease) => ({
        allowed: r.aliases,
        remember: true,
      }));
      setSecretConsent({ prompt });

      expect(await allowedSecrets(release())).toEqual(["gmail.imap"]);
      expect(await allowedSecrets(release())).toEqual(["gmail.imap"]);
      // Remembered nowhere, so asked both times — the safe direction.
      expect(prompt).toHaveBeenCalledTimes(2);
    } finally {
      if (real) {
        Object.defineProperty(window, "localStorage", real);
      }
    }
  });
});

File: meander/backend/tests/grants.test.cpp
#include <catch2/catch_test_macros.hpp>

#include <filesystem>
#include <fstream>
#include <string>

#include "../grants.h"

// ──────────────────────────────────────────────────────────────────────────────
// The remembered answers to "send secrets to this runtime?".
//
// A grant is not a credential, but it is what stands between a board and one:
// anything that can add a line here can have a secret handed over without being
// asked. So it is written like the vault — owner-only — and it is merged rather
// than replaced, because a board that comes to need one more secret is asked
// about that one alone and the answer must not drop what was already agreed.
// ──────────────────────────────────────────────────────────────────────────────

namespace {

/** Grants of their own, in a directory that does not outlive the test. */
class TempGrants
{
public:
  explicit TempGrants(const std::string& contents = "")
    : m_dir(std::filesystem::temp_directory_path() /
            ("readymade-grants-test-" + std::to_string(++s_counter)))
  {
    std::filesystem::create_directories(m_dir);
    if (!contents.empty())
    {
      std::ofstream file(path());
      file << contents;
    }
  }

  ~TempGrants()
  {
    std::error_code ec;
    std::filesystem::remove_all(m_dir, ec);
  }

  std::filesystem::path path() const { return m_dir / "grants.json"; }
  Grants open() const { return Grants(path()); }

private:
  std::filesystem::path m_dir;
  static inline int s_counter = 0;
};

const std::string MAIL_NODE = R"(["Mail","node","http://127.0.0.1:8080"])";
const std::string MAIL_OTHER = R"(["Mail","node","https://evil.example"])";

} // namespace

TEST_CASE("nothing is granted until something grants it", "[grants]") {
  TempGrants temp;
  Grants grants = temp.open();

  REQUIRE(grants.granted(MAIL_NODE).empty());
  REQUIRE(grants.getAll().empty());
}

TEST_CASE("what was granted is what comes back", "[grants]") {
  TempGrants temp;
  Grants grants = temp.open();

  REQUIRE(grants.grant(MAIL_NODE, {"gmail.imap"}));
  REQUIRE(grants.granted(MAIL_NODE) == std::vector<std::string>{"gmail.imap"});
}

TEST_CASE("a second grant adds to the first rather than replacing it", "[grants]") {
  // The case this guards: a board that comes to need `slack` is asked about
  // `slack` alone, so the answer carries only that — and must not be read as
  // withdrawing the secret already agreed to.
  TempGrants temp;
  Grants grants = temp.open();
  grants.grant(MAIL_NODE, {"gmail.imap"});

  REQUIRE(grants.grant(MAIL_NODE, {"slack"}));
  REQUIRE(grants.granted(MAIL_NODE) ==
          std::vector<std::string>{"gmail.imap", "slack"});
}

TEST_CASE("granting the same alias twice changes nothing", "[grants]") {
  TempGrants temp;
  Grants grants = temp.open();
  grants.grant(MAIL_NODE, {"gmail.imap"});
  grants.grant(MAIL_NODE, {"gmail.imap"});

  REQUIRE(grants.granted(MAIL_NODE) == std::vector<std::string>{"gmail.imap"});
}

TEST_CASE("a grant is bound to the address it was given for", "[grants]") {
  // A runtime id is board-controlled and means nothing alone: the same `node`
  // is repointed at another server by editing one field, and the grant must not
  // follow it there.
  TempGrants temp;
  Grants grants = temp.open();
  grants.grant(MAIL_NODE, {"gmail.imap"});

  REQUIRE(grants.granted(MAIL_OTHER).empty());
}

TEST_CASE("revoking one makes the next release ask again", "[grants]") {
  TempGrants temp;
  Grants grants = temp.open();
  grants.grant(MAIL_NODE, {"gmail.imap"});

  REQUIRE(grants.revoke(MAIL_NODE));
  REQUIRE(grants.granted(MAIL_NODE).empty());
}

TEST_CASE("revoking one that was never given is not a failure to report", "[grants]") {
  TempGrants temp;
  Grants grants = temp.open();

  REQUIRE_FALSE(grants.revoke(MAIL_NODE));
}

TEST_CASE("granting nothing grants nothing", "[grants]") {
  // A refusal reaches here as an empty list. Writing an empty entry would read
  // back as a key that exists, which is not the same as having agreed to it.
  TempGrants temp;
  Grants grants = temp.open();

  REQUIRE_FALSE(grants.grant(MAIL_NODE, {}));
  REQUIRE(grants.getAll().empty());
}

TEST_CASE("a key holding anything but names is ignored", "[grants]") {
  TempGrants temp(R"({"a":"not a list","b":[1,2],"c":["real"]})");
  Grants grants = temp.open();

  REQUIRE(grants.granted("a").empty());
  REQUIRE(grants.granted("b").empty());
  REQUIRE(grants.granted("c") == std::vector<std::string>{"real"});
  REQUIRE(grants.getAll().size() == 1);
}

TEST_CASE("a board named like a separator is still its own board", "[grants]") {
  // Why the key is a JSON array and not joined text: a board may be called
  // anything, and two different triples must never read as one.
  TempGrants temp;
  Grants grants = temp.open();
  grants.grant(R"(["Mail node","x","http://a"])", {"one"});
  grants.grant(R"(["Mail","node x","http://a"])", {"two"});

  REQUIRE(grants.granted(R"(["Mail node","x","http://a"])") ==
          std::vector<std::string>{"one"});
  REQUIRE(grants.granted(R"(["Mail","node x","http://a"])") ==
          std::vector<std::string>{"two"});
}

TEST_CASE("malformed json is no grants rather than a crash", "[grants]") {
  TempGrants temp("{not json at all");
  Grants grants = temp.open();

  REQUIRE(grants.getAll().empty());
  REQUIRE(grants.granted(MAIL_NODE).empty());
}

TEST_CASE("listing walks grants that outlive the call", "[grants]") {
  // The same dangling-temporary trap `vault.h` had: `items()` refers into the
  // json rather than copying it, so iterating one straight off a read walks a
  // container destroyed at the semicolon.
  TempGrants temp(R"({"a":["one"],"b":["two"],"c":["three"]})");
  Grants grants = temp.open();

  REQUIRE(grants.getAll().size() == 3);
}

TEST_CASE("the file is readable only by its owner", "[grants]") {
  TempGrants temp;
  Grants grants = temp.open();
  grants.grant(MAIL_NODE, {"gmail.imap"});

  const auto mode = std::filesystem::status(temp.path()).permissions();
  REQUIRE((mode & std::filesystem::perms::group_all) == std::filesystem::perms::none);
  REQUIRE((mode & std::filesystem::perms::others_all) == std::filesystem::perms::none);
}
